lib/tooling/base-tool.ts

```diff
@@ -60,81 +68,88 @@ export class BaseTool {
+            maxErrorOutput: this.env.ceProps('max-error-output', 5000) as number,
             wrapper: this.env.compilerProps('compiler-wrapper'),
         };
     }
 
-    createErrorResponse(message) {
+    // By default calls utils.parseOutput, but lets subclasses override their output processing
+    protected parseOutput(lines: string, inputFilename?: string, pathPrefix?: string): ResultLine[] {
+        return parseOutput(lines, inputFilename, pathPrefix);
+    }
+
+    createErrorResponse(message: string): ToolResult {
         return {
             id: this.tool.id,
             name: this.tool.name,
             code: -1,
             languageId: 'stderr',
             stdout: [],
-            stderr: utils.parseOutput(message),
+            stderr: this.parseOutput(message),
         };
     }
 
     // mostly copy&paste from base-compiler.js
-    findLibVersion(selectedLib, supportedLibraries) {
+    findLibVersion(selectedLib: SelectedLibraryVersion, supportedLibraries: Record<string, Library>) {
         const foundLib = _.find(supportedLibraries, (o, libId) => libId === selectedLib.id);
         if (!foundLib) return false;
 
         return _.find(foundLib.versions, (o, versionId) => versionId === selectedLib.version);
     }
 
     // mostly copy&paste from base-compiler.js
-    getIncludeArguments(libraries, supportedLibraries) {
+    getIncludeArguments(libraries: SelectedLibraryVersion[], supportedLibraries: Record<string, Library>): string[] {
         const includeFlag = '-I';
 
-        return _.flatten(
-            _.map(libraries, selectedLib => {
-                const foundVersion = this.findLibVersion(selectedLib, supportedLibraries);
-                if (!foundVersion) return false;
+        return libraries.flatMap(selectedLib => {
+            const foundVersion = this.findLibVersion(selectedLib, supportedLibraries);
+            if (!foundVersion) return [];
 
-                return _.map(foundVersion.path, path => includeFlag + path);
-            }),
-        );
+            return foundVersion.path.map(path => includeFlag + path);
+        });
     }
 
-    getLibraryOptions(libraries, supportedLibraries) {
-        return _.flatten(
-            _.map(libraries, selectedLib => {
-                const foundVersion = this.findLibVersion(selectedLib, supportedLibraries);
-                if (!foundVersion) return false;
+    getLibraryOptions(libraries: SelectedLibraryVersion[], supportedLibraries: Record<string, Library>): string[] {
+        return libraries.flatMap(selectedLib => {
+            const foundVersion = this.findLibVersion(selectedLib, supportedLibraries);
+            if (!foundVersion) return [];
 
-                return foundVersion.options;
-            }),
-        );
+            return foundVersion.options;
+        });
     }
 
-    async runTool(compilationInfo, inputFilepath, args, stdin /*, supportedLibraries*/) {
+    async runTool(
+        compilationInfo: Record<any, any>,
```

Comment:
Needs a more specific type, but I wasn't sure which

Reply:
This is whatever getCompilationInfo returns

Reply:
yes, but as that function extends one of its parameters, and that one does not have type info either, I'm not too sure what the best thing to do is